### ISSUE TITLE
iOS crash on exit (CPUParticles2D::_set_redraw is called on NOTIFICATION_EXIT_TREE)

### ISSUE THREAD

**Issue by naithar** [godotengine/godot#39786]
**Godot version:**

Godot 3.2.1/3.2.2 rc

**OS/device including version:**

iOS 13.5 both GLES2 and GLES3

**Issue description:**

Upon closing application without sending it to background, application crashes with this stack trace:
```
#0	ThreadSafe::lock() const at core/os/thread_safe.h:42
#1	ThreadSafeMethod::ThreadSafeMethod(ThreadSafe const*) at core/os/thread_safe.h:60
#2	ThreadSafeMethod::ThreadSafeMethod(ThreadSafe const*) at godot/core/os/thread_safe.h:57
#3	MessageQueue::push_call(unsigned long long, StringName const&, Variant const**, int, bool) at core/message_queue.cpp:47
#4	MessageQueue::push_call(unsigned long long, StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) at core/message_queue.cpp:92
#5	MessageQueue::push_call(Object*, StringName const&, Variant const&, Variant const&, Variant const&, Variant const&, Variant const&) at core/message_queue.cpp:153
#6	CanvasItem::update() atscene/2d/canvas_item.cpp:642
#7	CPUParticles2D::_set_redraw(bool) at scene/2d/cpu_particles_2d.cpp:1073
#8	CPUParticles2D::_notification(int) at scene/2d/cpu_particles_2d.cpp:1099
#9	CPUParticles2D::_notificationv(int, bool) at scene/2d/cpu_particles_2d.h:40
#10	Object::notification(int, bool) at core/object.cpp:932
#11	Node::_propagate_exit_tree() at scene/main/node.cpp:305
#12	Node::_propagate_exit_tree() at scene/main/node.cpp:294
#13	Node::_propagate_exit_tree() at scene/main/node.cpp:294
#14	Node::_set_tree(SceneTree*) at scene/main/node.cpp:2560
#15	SceneTree::finish() at scene/main/scene_tree.cpp:619
#16	OSIPhone::delete_main_loop() at platform/iphone/os_iphone.cpp:350
#17	Main::cleanup() at main/main.cpp:2233
#18	iphone_finish() at platform/iphone/godot_iphone.cpp:90
```

**Steps to reproduce:**
Add `CPUParticles2D` to node tree, export to iOS, run and close application by swiping without going to home or moving it to background.

**Minimal reproduction project:**

https://github.com/naithar/gd_richedit_test/tree/crash_on_close

**Comment by naithar** [comment]
After looking into the issue I've found out that crash happens because when `CPUParticles2D` calls `update` method after `Main::cleanup()` it sends `_update_callback` call to `MessageQueue` while it's `get_singleton()` returns `null`. So it could possibly affect other platforms.

**Comment by naithar** [comment]
Reverting changes in this commit https://github.com/godotengine/godot/commit/ff59c562f745de957e8395fc02af1e5d7377a302 seems to be fixing this issue, but I'm not sure if it will not return the crashing of editor that this commit fixed as I was unable to reproduce the crash in editor.

**Comment by naithar** [comment]
Adding a `Application can be killed immediately when user is shutting down or logging out` or `NSSupportsSuddenTermination` key with value `false` into Info.plist file make this crash appear even if app is closed while in background.

**Comment by akien-mga** [comment]
@naithar Thanks for the great debugging work! Could you check if #39798 fixes it for you? (same as reverting ff59c56 but also keeps an early `flush()` where it was until now).

**Comment by naithar** [comment]
@akien-mga this was actually the first thing that I did, too check if the problem was in freeing `MessageQueue` and it's working :) 
But I couldn't reproduce the crash of the editor that should have been fixed, so I'm not sure if this fix will not break on any other platform. Though I've tested fix on multiple devices and systems and I've found no problem with it.
